sacc2/sacc.py

```diff
@@ -0,0 +1,871 @@
+        """
+        return self._get_2pt(data_type, bin1, bin2, return_cov, 'theta')
+
+    def _add_2pt(self, data_type, bin1, bin2, x, tag_val, tag_name, window):
+        """
+        Internal method for adding 2pt data points.
+        Copes with multiple values for the parameters
+        """
+        # single data point case
+        if np.isscalar(tag_val):
+            t = {tag_name: float(tag_val)}
+            if window is not None:
+                t['window'] = window
+            self.add_data_point(data_type, (bin1, bin2), x, **t)
+            return
+        # multiple ell/theta values but same bin
+        elif np.isscalar(bin1):
+            n1 = len(x)
+            n2 = len(tag_val)
+            if not n1 == n2:
+                raise ValueError(f"Length of inputs do not match in added 2pt data ({n1},{n2})")
+            if window is None:
+                for tag_i, x_i in zip(tag_val, x):
+                    self._add_2pt(data_type, bin1, bin2, x_i, tag_i, tag_name, window)
+            else:
+                for tag_i, x_i, w_i in zip(tag_val, x, window):
+                    self._add_2pt(data_type, bin1, bin2, x_i, tag_i, tag_name, w_i)
+        # multiple bin values
+        elif np.isscalar(data_type):
+            n1 = len(x)
+            n2 = len(tag_val)
+            n3 = len(bin1)
+            n4 = len(bin2)
+            if not (n1 == n2 == n3 == n4):
+                raise ValueError(f"Length of inputs do not match in added 2pt data ({n1}, {n2}, {n3}, {n4})")
+            if window is None:
+                for b1, b2, tag_i, x_i in zip(bin1, bin2, tag_val, x):
+                    self._add_2pt(data_type, b1, b2, x_i, tag_i, tag_name, window)
+            else:
+                for b1, b2, tag_i, x_i, w_i in zip(bin1, bin2, tag_val, x, window):
+                    self._add_2pt(data_type, b1, x_i, tag_i, tag_name, w_i)
+        # multiple data point values
+        else:
+            n1 = len(x)
+            n2 = len(tag_val)
+            n3 = len(bin1)
+            n4 = len(bin2)
+            n5 = len(data_type)
+            if not (n1 == n2 == n3 == n4 == n5):
+                raise ValueError(f"Length of inputs do not match in added 2pt data ({n1}, {n2}, {n3}, {n4}, {n5})")
+            if window is None:
+                for d, b1, b2, tag_i, x_i in zip(data_type, bin1, bin2, tag_val, x):
+                    self._add_2pt(d, b1, b2, x_i, tag_i, tag_name, window)
+            else:
+                for d, b1, b2, tag_i, x_i, w_i in zip(data_type, bin1, bin2, tag_val, x, window):
+                    self._add_2pt(d, b1, b2, x_i, tag_i, tag_name, w_i)
+
+    def add_ell_cl(self, data_type, bin1, bin2, ell, x, window=None):
+        """
+        Add a series of 2pt Fourier space data points, either
+        individually or as a group.
+
+        data_type: str or array/list of str
+            Which type C_ell to add
+        bin1: str or array/list of str
+            The name(s) of the first tomographic bin
```

Comment:
the mention of tomographic bins here seem too specific to me. We want this to be used for other things (cmb lensing, etc.). In order to keep the notation homogeneous with the rest of the code, I'd change the docstrings and rename bin1, bin2 to tracer1, tracer2. Same for `_add_2pt`, `add_theta_xi`, and the corresponding `get` functions.

Reply:
These are convenience functions for 3x2pt, where the tracer does correspond to a tomographic bin.  Do you think that the name change is confusing?

Reply:
OK, but the function would be exactly the same if you wanted to retrieve power spectra between CMB kappa and delta_g, right? Or between any other pair of maps. So it'd make sense to have one single function for this, rather than copies of the same function with different documentation, right?

I think this brings me back to the interpretation of data_type as something that is separate from the physical properties of the tracers you're correlating (the argument is the same there: you don't need to define different types for power spectra between different types of tracers)

Reply:
That's a good point - I'll switch to tracer1, tracer2 and put a note in the docs that tomographic bins are one example.